diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,289 @@
+                # correct one for each dimension
+                return lax.cond(
+                    jnp.all(cond), lambda _: tay_out, select_tay_or_direct, None
+                )
+
+        tree_with_cfs = jtu.tree_map(recompute_coeffs_leaf, gamma, tay_cfs)
+        outer = jtu.tree_structure(gamma)
+        inner = self._coeffs_structure
+        cfs_with_tree = jtu.tree_transpose(outer, inner, tree_with_cfs)
+        return cfs_with_tree
+
+    def init(
+        self,
+        terms: LangevinTerm,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: LangevinTuple,
+        args: PyTree,
+    ) -> _SolverState:
+        """Precompute _SolverState which carries the Taylor coefficients and the
+        ALIGN coefficients (which can be computed from h and the Taylor coeffs).
+        This method is FSAL, so _SolverState also carries the previous evaluation
+        of grad_f.
+        """
+        assert isinstance(terms, LangevinTerm)
+        gamma, u, f = terms.args  # f is in fact grad(f)
+
+        x0, v0 = y0
+
+        def _check_shapes(_c, _u, _x, _v):
+            # assert _x.ndim in [0, 1]
+            assert _c.shape == _u.shape == _x.shape == _v.shape
+
+        assert jtu.tree_all(jtu.tree_map(_check_shapes, gamma, u, x0, v0))
+
+        h = t1 - t0
+
+        tay_cfs = jtu.tree_map(self._comp_taylor_coeffs_leaf, gamma)
+        # tay_cfs have the same tree structure as gamma, with each leaf being a _Coeffs
+        # and the arrays have an extra trailing dimension of 6
+
+        coeffs = self._recompute_coeffs(h, gamma, tay_cfs)
+        rho = jtu.tree_map(lambda c, _u: jnp.sqrt(2 * c * _u), gamma, u)
+
+        state_out = _SolverState(
+            h=h,
+            taylor_coeffs=tay_cfs,
+            coeffs=coeffs,
+            rho=rho,
+            prev_f=f(x0),
+        )
+
+        return state_out
+
+    @staticmethod
+    @abc.abstractmethod
+    def _compute_step(
+        h: RealScalarLike,
+        levy,
+        x0: LangevinX,
+        v0: LangevinX,
+        langevin_args: _LangevinArgs,
+        cfs: _Coeffs,
+        st: _SolverState,
+    ) -> tuple[LangevinX, LangevinX, LangevinX, _ErrorEstimate]:
+        raise NotImplementedError
+
+    def step(
+        self,
+        terms: LangevinTerm,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: LangevinTuple,
+        args: PyTree,
+        solver_state: _SolverState,
+        made_jump: BoolScalarLike,
+    ) -> tuple[LangevinTuple, _ErrorEstimate, DenseInfo, _SolverState, RESULTS]:
+        del made_jump, args
+        st = solver_state
+        h = t1 - t0
+        assert isinstance(terms, LangevinTerm)
+        gamma, u, f = terms.args
+
+        h_state = st.h
+        tay: PyTree[_Coeffs] = st.taylor_coeffs
+        cfs: _Coeffs = st.coeffs  # type: ignore
+
+        # If h changed recompute coefficients
+        cond = jnp.isclose(h_state, h)
```

Comment:
I'd suggest perhaps making this equality? This seems like we're leaving numerical correctness kind of up to chance here.

Reply:
Maybe I can make the tolerance smaller, but for some reason requiring equality would make the coeffs be recomputed way too often (although it's been a while since I experimented with this, so I could be wrong).

Reply:
So yes I redid the experiments and even with a constant step controller, the step sizes fluctuate by cca 1e-16, which is negligible, but if I have `==` instead of `jnp.isclose`, then the coefficients get recomputed way too often. So I'll stick with `cond = jnp.isclose(h_state, h, rtol=1e-10, atol=1e-12)`.